src/liballoc/rc.rs

```diff
@@ -1708,9 +1704,18 @@ impl<T> Weak<T> {
     /// [`null`]: ../../std/ptr/fn.null.html
     #[stable(feature = "weak_into_raw", since = "1.45.0")]
     pub fn as_ptr(&self) -> *const T {
-        let offset = data_offset_sized::<T>();
-        let ptr = self.ptr.cast::<u8>().as_ptr().wrapping_offset(offset);
-        ptr as *const T
+        let ptr: *mut RcBox<T> = NonNull::as_ptr(self.ptr);
+        let fake_ptr = ptr as *mut T;
+
+        // SAFETY: we must offset the pointer manually, and said pointer may be
+        // a dangling weak (usize::MAX). data_offset is safe to call, because we
```

Comment:
Is `ptr` the pointer here?

`(*ptr).value` is still UB when `ptr` dangles, no matter the `&raw const` around it.

Reply:
You're right, of course. That's why we're doing the offset manually here in the first place.

Thankfully, we can just use `fake_ptr` here rather than `&raw const (*ptr).value)`, since the argument to `data_offset` is only required to have the correct pointer metadata (here acquired via `as` cast from the stored pointer) and not to point to an actual `T`.

Reply:
Honestly, I'm not quite happy with this function and the use of `fake_ptr`; I think it obscures the intent. Do you think the following is better?

```rust
    pub fn as_ptr(&self) -> *const T {
        let ptr: *mut RcBox<T> = NonNull::as_ptr(self.ptr);

        // SAFETY: we must offset the pointer manually, and said pointer may be
        // a dangling weak (usize::MAX). data_offset is safe to call, because we
        // know a pointer to unsized T must be derived from a real unsized T,
        // because dangling weaks are only created for sized T. wrapping_offset
        // is used so that we can use the same code path for dangling weak refs.
        unsafe {
            let offset = data_offset(ptr as *mut T);
            set_data_ptr(ptr as *mut T, (ptr as *mut u8).wrapping_offset(offset))
        }
    }
```

Reply:
```
 // SAFETY: we must offset the pointer manually, and said pointer may be
  // a dangling weak (usize::MAX).
```
I would add here already "but only for sized `T`" or so. Currently you are introducing that restriction later, I think it is best to state all such things derived from `Weak`'s invariant up-front.

What do you mean by "inlining `fake_ptr`"?

Reply:
Oh, `fake_ptr` is a let-bound variable that you inline... I somehow thought you meant a function.^^